modules/table/src/main/java/org/apache/ignite/internal/table/distributed/gc/MvGc.java

```diff
@@ -0,0 +1,249 @@
+     *
+     * @param newLwm New low watermark.
+     * @throws IgniteInternalException with {@link Gc#CLOSED_ERR} If the garbage collector is closed.
+     */
+    public void updateLowWatermark(HybridTimestamp newLwm) {
+        inBusyLock(() -> {
+            HybridTimestamp updatedLwm = lowWatermarkReference.updateAndGet(currentLwm -> {
+                if (currentLwm == null) {
+                    return newLwm;
+                }
+
+                // Update only if the new one is greater than the current one.
+                return newLwm.compareTo(currentLwm) > 0 ? newLwm : currentLwm;
+            });
+
+            // If the new watermark is smaller than the current one or has been updated in parallel, then we do nothing.
+            if (updatedLwm != newLwm) {
+                return;
+            }
+
+            executor.submit(() -> inBusyLock(this::initNewGcBusy));
+        });
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        shutdownAndAwaitTermination(executor, 10, TimeUnit.SECONDS);
+    }
+
+    private void initNewGcBusy() {
+        storageHandlerByPartitionId.keySet().forEach(this::scheduleGcForStorage);
+    }
+
+    private void scheduleGcForStorage(TablePartitionId tablePartitionId) {
+        executor.submit(() -> inBusyLock(() -> {
+            GcStorageHandler storageHandler = storageHandlerByPartitionId.compute(tablePartitionId, (id, gcStorageHandler) -> {
+                if (gcStorageHandler == null) {
+                    // Storage has been removed from garbage collection.
+                    return gcStorageHandler;
+                }
+
+                boolean casResult = gcStorageHandler.gcInProgressFuture.compareAndSet(null, new CompletableFuture<>());
+
+                assert casResult : tablePartitionId;
+
+                return gcStorageHandler;
+            });
+
+            if (storageHandler == null) {
+                // Storage has been removed from garbage collection.
+                return;
+            }
+
+            CompletableFuture<Void> future = storageHandler.gcInProgressFuture.get();
+
+            assert future != null : tablePartitionId;
+
+            try {
+                boolean scheduleGcForStorageAgain = true;
+
+                for (int i = 0; i < GC_BUTCH_SIZE && scheduleGcForStorageAgain; i++) {
+                    HybridTimestamp lowWatermark = lowWatermarkReference.get();
+
+                    assert lowWatermark != null : tablePartitionId;
+
+                    // If storage has been deleted or there is no garbage, then for now we will stop collecting garbage for this storage.
+                    if (!storageHandlerByPartitionId.containsKey(tablePartitionId)
+                            || !storageHandler.storageUpdateHandler.vacuum(lowWatermark)) {
+                        scheduleGcForStorageAgain = false;
```

Comment:
You can just return from the method right away instead of having an additional flag. The less variables you have, the simpler your code is, generally speaking.
And by the way, future can be completed in finally block, another way to simplify the code. Just don't forget that it can already be completed with exception by that time

Reply:
Fix it.